core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -95,20 +95,66 @@ private[spark] class SparkSubmit extends Logging {
   }
 
   /**
-   * Kill an existing submission using the REST protocol. Standalone and Mesos cluster mode only.
+   * Kill an existing submission.
+   * Standalone, Kubernetes and Mesos cluster mode only.
    */
   private def kill(args: SparkSubmitArguments): Unit = {
-    new RestSubmissionClient(args.master)
-      .killSubmission(args.submissionToKill)
+    if (args.master.startsWith("k8s://")) {
+      try {
+        val ops = Utils.classForName("org.apache.spark.deploy.k8s.submit.K8sSubmitOps$")
```

Comment:
There are two approaches one that uses a rest server/client and the other that uses whatever the env provides. For k8s there is no rest api to use directly. We communicate to the k8s api server via the fabricio client lib later on. Reflection is needed because we need to access classes which are not by default available in every Spark build. For example if a build does not target k8s then these classes will not be available. Also Spark-submit code is independent of the resource manager dependencies so we should reference in there any types that come from the K8s backend.

Reply:
I understand why the code needs to be different, but I agree with the sentiment that adding this kind of code to `SparkSubmit` is what we should be trying to avoid at this point.

This is the kind of thing that belongs behind an interface that each RM backend should then implement, like `ExternalClusterManager` does for the scheduler. So if you have an interface e.g. `SparkSubmitBackend` which k8s then implements, this code can all be encapsulated in the k8s module, can be more easily tested, and moves `SparkSubmit` in the right direction, where all RM-specific functionality would eventually be behind that interface (instead of the current mess).

Reply:
@vanzin 
If you have an interface then you need to initialize an instance out of it. That means you need to use some form of class loading. Spark context is initializing the right cluster manager by using:
```
  private def getClusterManager(url: String): Option[ExternalClusterManager] = {
    val loader = Utils.getContextOrSparkClassLoader
    val serviceLoaders =
      ServiceLoader.load(classOf[ExternalClusterManager], loader).asScala.filter(_.canCreate(url))
....
```
and picks up the META-INF/Services defined at build time. 
Are you implying that I should use something like that?

Reply:
> Do you mean that I should use something like that? 

Yes. That would make submission pluggable like the scheduler is, using a similar mechanism. Eventually I hope we can hide all of the details that SparkSubmit handles behind that interface...

Reply:
Ok @vanzin could you also have a look at https://github.com/apache/spark/pull/23546. Although a bit different case, also affects Spark submit.

Reply:
I think parts of that change actually pose a similar issue. Basically, the part that deals with how to stage dependencies for cluster-mode submissions, since each backend may do something slightly different (like YARN does its own thing, even if the "copy files to HDFS" part is somewhat similar).

But in this particular case, this big chunk of reflection code is begging to just be turned into a plugin. You don't need fully hash out that plugin interface, but at least add the basics of what you're trying to achieve in this PR to it. We can look at adding more things later as needed.